### ISSUE TITLE
ValueError: max() arg is an empty sequence

### ISSUE THREAD

**Issue by mhooykaas** [whatshap/whatshap#439]
Hello,

While trying to run WhatsHap 1.7 polyphasegenetic (python 3.10.8) I sometimes get a ValueError. In this case I ran the same vcf file with different --complexity-support values and it only resulted in this error with --complexity-support 2. However on other VCFs I also saw the error with the default setting of --complexity-support (SxN markers only)

I installed WhatsHap with conda.

This command was succesfully run:
`time whatshap polyphasegenetic ${VCF} ${PEDIGREE} --ploidy ${PLOIDY} --sample Altus_470bp --sample Colomba_470bp --distrust-genotypes -o ${OUTVCF} --complexity-support 1`

This resulted in an error:
`time whatshap polyphasegenetic ${VCF} ${PEDIGREE} --ploidy ${PLOIDY} --sample Altus_470bp --sample Colomba_470bp --distrust-genotypes -o ${OUTVCF} --complexity-support 2`

This was in the log file:
```
This is WhatsHap (polyploid-genetic) 1.7 running under Python 3.10.8
======== Working on chromosome 'StSOLv1.1ch03'
Number of variants among all samples: 6481
---- Processing individual Colomba_470bp
Detected Altus_470bp as co-parent for Colomba_470bp.
Extracting progeny allele depths ...
Computing progeny genotype likelihoods ...
   Number of nodes to cluster: 1664
   Number of simplex-nulliplex variants: 868
   Correcting variant type based on progenies:
   1/0 (868)
      -> 1/0: 774 (89.2%)
      -> 0/0: 78 (9.0%)
      -> 2/0: 16 (1.8%)
   1/1 (476)
      -> 2/0: 216 (45.4%)
      -> 1/0: 179 (37.6%)
      -> 0/0: 43 (9.0%)
      -> 3/0: 31 (6.5%)
      -> 2/1: 3 (0.6%)
      -> 1/1: 1 (0.2%)
      -> 4/0: 3 (0.6%)
   2/0 (160)
      -> 2/0: 142 (88.8%)
      -> 2/1: 2 (1.2%)
      -> 1/0: 15 (9.4%)
      -> 3/0: 1 (0.6%)
   Number of nodes to cluster: 1106
   Number of simplex-nulliplex variants: 789
Compute scores for markers ...
Clustering marker alleles ...
Arranging clusters ...
Traceback (most recent call last):
  File "/media/bulk_01/users/hooyk002/miniconda3/envs/whatshap/bin/whatshap", line 10, in <module>
    sys.exit(main())
  File "/media/bulk_01/users/hooyk002/miniconda3/envs/whatshap/lib/python3.10/site-packages/whatshap/__main__.py", line 64, in main
    module.main(args)
  File "/media/bulk_01/users/hooyk002/miniconda3/envs/whatshap/lib/python3.10/site-packages/whatshap/cli/polyphasegenetic.py", line 607, in main
    run_polyphasegenetic(**vars(args))
  File "/media/bulk_01/users/hooyk002/miniconda3/envs/whatshap/lib/python3.10/site-packages/whatshap/cli/polyphasegenetic.py", line 192, in run_polyphasegenetic
    superreads[sample], components[sample] = phase_single_sample(
  File "/media/bulk_01/users/hooyk002/miniconda3/envs/whatshap/lib/python3.10/site-packages/whatshap/cli/polyphasegenetic.py", line 301, in phase_single_sample
    haplo_skeletons = arrange_clusters(clustering, padding, param.ploidy)
  File "/media/bulk_01/users/hooyk002/miniconda3/envs/whatshap/lib/python3.10/site-packages/whatshap/polyphase/clusterarrangement.py", line 34, in arrange_clusters
    n = max(c_end)
ValueError: max() arg is an empty sequence

real    0m27.605s
user    0m27.883s
sys     0m8.584s
```

Thanks in advance!

**Comment by schrins** [comment]
It looks like there is no cluster of markers with at least two markers (or no cluster at all). It's inconvenient that the code crashes at this point for such a case, but the more interesting question is probably why there are no such clusters. Was the above output the full output? There should be some numeric stuff in between the different steps (see example below). If at some point there zero variants/markers in an intermediate step, this would explain, why the code crashes at the place you reported. If not, I need to dig deeper what is going on.

```
======== Working on chromosome 'StSOLv1.1ch04'
Number of variants among all samples: 16777
---- Processing individual Altus_470bp
Detected Colomba_470bp as co-parent for Altus_470bp.
Extracting progeny allele depths ...
Computing progeny genotype likelihoods ...
   Number of nodes to cluster: 3600
   Number of simplex-nulliplex variants: 3600
   Correcting variant type based on progenies:
   1/0 (3600)
      -> 1/0: 3156 (87.7%)
      -> 0/0: 422 (11.7%)
      -> 2/0: 22 (0.6%)
   Number of nodes to cluster: 3200
   Number of simplex-nulliplex variants: 3156
Compute scores for markers ...
Clustering marker alleles ...
Arranging clusters ...
Arranged 3071 variants out of a total of 3200
   h0: [10]
   h1: [1]
   h2: [2]
   h3: [0]
---- Processing individual Colomba_470bp
Detected Altus_470bp as co-parent for Colomba_470bp.
Extracting progeny allele depths ...
Computing progeny genotype likelihoods ...
   Number of nodes to cluster: 2528
   Number of simplex-nulliplex variants: 2528
   Correcting variant type based on progenies:
   1/0 (2528)
      -> 1/0: 1973 (78.0%)
      -> 0/0: 532 (21.0%)
      -> 2/0: 23 (0.9%)
   Number of nodes to cluster: 2019
   Number of simplex-nulliplex variants: 1973
Compute scores for markers ...
Clustering marker alleles ...
Arranging clusters ...
Arranged 1868 variants out of a total of 2019
   h0: [0]
   h1: [1]
   h2: [4]
   h3: [8]
======== Writing VCF
Done writing VCF
```

**Comment by mhooykaas** [comment]
Thank you for your response. There were markers of all three types. I have updated my post above to include the entire output. That was on a vcf with part of the markers of chr03. When I ran it on all markers it also resulted in this error:
```
This is WhatsHap (polyploid-genetic) 1.7 running under Python 3.10.8
======== Working on chromosome 'StSOLv1.1ch03'
Number of variants among all samples: 2145261
---- Processing individual Colomba_470bp
Detected Altus_470bp as co-parent for Colomba_470bp.
Extracting progeny allele depths ...
Computing progeny genotype likelihoods ...
   Number of nodes to cluster: 192243
   Number of simplex-nulliplex variants: 192243
   Correcting variant type based on progenies:
   1/0 (192243)
      -> 1/0: 154436 (80.3%)
      -> 2/0: 4959 (2.6%)
      -> 0/0: 32848 (17.1%)
   Number of nodes to cluster: 164354
   Number of simplex-nulliplex variants: 154436
Compute scores for markers ...
Clustering marker alleles ...
Arranging clusters ...
Traceback (most recent call last):
  File "/media/bulk_01/users/hooyk002/miniconda3/envs/whatshap/bin/whatshap", line 10, in <module>
    sys.exit(main())
  File "/media/bulk_01/users/hooyk002/miniconda3/envs/whatshap/lib/python3.10/site-packages/whatshap/__main__.py", line 64, in main
    module.main(args)
  File "/media/bulk_01/users/hooyk002/miniconda3/envs/whatshap/lib/python3.10/site-packages/whatshap/cli/polyphasegenetic.py", line 607, in main
    run_polyphasegenetic(**vars(args))
  File "/media/bulk_01/users/hooyk002/miniconda3/envs/whatshap/lib/python3.10/site-packages/whatshap/cli/polyphasegenetic.py", line 192, in run_polyphasegenetic
    superreads[sample], components[sample] = phase_single_sample(
  File "/media/bulk_01/users/hooyk002/miniconda3/envs/whatshap/lib/python3.10/site-packages/whatshap/cli/polyphasegenetic.py", line 301, in phase_single_sample
    haplo_skeletons = arrange_clusters(clustering, padding, param.ploidy)
  File "/media/bulk_01/users/hooyk002/miniconda3/envs/whatshap/lib/python3.10/site-packages/whatshap/polyphase/clusterarrangement.py", line 34, in arrange_clusters
    n = max(c_end)
ValueError: max() arg is an empty sequence
```


Below is the output when I ran WhatsHap on the same vcf, but with --complexity-support 1 instead of 2 and there was no error:

````
This is WhatsHap (polyploid-genetic) 1.7 running under Python 3.10.8
======== Working on chromosome 'StSOLv1.1ch03'
Number of variants among all samples: 6481
---- Processing individual Altus_470bp
Detected Colomba_470bp as co-parent for Altus_470bp.
Extracting progeny allele depths ...
Computing progeny genotype likelihoods ...
   Number of nodes to cluster: 2621
   Number of simplex-nulliplex variants: 2145
   Correcting variant type based on progenies:
   1/0 (2145)
      -> 1/0: 1947 (90.8%)
      -> 0/0: 156 (7.3%)
      -> 2/0: 42 (2.0%)
   1/1 (476)
      -> 2/0: 216 (45.4%)
      -> 1/0: 179 (37.6%)
      -> 0/0: 43 (9.0%)
      -> 3/0: 31 (6.5%)
      -> 2/1: 3 (0.6%)
      -> 1/1: 1 (0.2%)
      -> 4/0: 3 (0.6%)
   Number of nodes to cluster: 2032
   Number of simplex-nulliplex variants: 1947
Compute scores for markers ...
Clustering marker alleles ...
Arranging clusters ...
Arranged 1818 variants out of a total of 2032
   h0: [4]
   h1: [19]
   h2: [15]
   h3: [5]
---- Processing individual Colomba_470bp
Detected Altus_470bp as co-parent for Colomba_470bp.
Extracting progeny allele depths ...
Computing progeny genotype likelihoods ...
   Number of nodes to cluster: 1344
   Number of simplex-nulliplex variants: 868
   Correcting variant type based on progenies:
   1/0 (868)
      -> 1/0: 774 (89.2%)
      -> 0/0: 78 (9.0%)
      -> 2/0: 16 (1.8%)
   1/1 (476)
      -> 2/0: 216 (45.4%)
      -> 1/0: 179 (37.6%)
      -> 0/0: 43 (9.0%)
      -> 3/0: 31 (6.5%)
      -> 2/1: 3 (0.6%)
      -> 1/1: 1 (0.2%)
      -> 4/0: 3 (0.6%)
   Number of nodes to cluster: 807
   Number of simplex-nulliplex variants: 774
Compute scores for markers ...
Clustering marker alleles ...
Arranging clusters ...
Arranged 716 variants out of a total of 807
   h0: [0]
   h1: [3]
   h2: [5]
   h3: [19]
======== Writing VCF
Done writing VCF

== SUMMARY ==
Maximum memory usage: 0.306 GB
Time spent parsing VCF:                      7.9 s
Time spent for genetic scoring:             42.1 s
Time spent for clustering:                   6.0 s
Time spent for cluster arrangement:          0.4 s
Time spent writing VCF:                      2.1 s
Time spent on rest:                          4.3 s
Total elapsed time:                         62.6 s

real    1m4.457s
user    1m5.664s
sys     0m24.057s
```

I noticed that the vcf files you uploaded to Zenodo had 'decluttered' in the file name. Do you preprocess vcf files in some way?

**Comment by schrins** [comment]
Yes, at some point I removed all information that is irrelevant for Whatshap (and also not touched by Whatshap) to have more human-readable files:
1. Remove everything in the INFO field
2. Remove everything from the FORMAT field except GT and AD
3. Remove all variants that are homozygous on Altus_470bp

I will try to reproduce the issue with the unchanged data, but this will likely not happen before Wednesday.

One thing I noticed though: Does your VCF file contain both parent and progeny information? I always tested with separated files, where you then have to specify the progeny file explicitly with `-P`. If you have everything in one file, you could try to specify the same file a second time with the `-P` flag. This should of course not make a difference, but since I stopped testing the single-file case at some point, chances are high, that there could be an undetected bug.

**Comment by mhooykaas** [comment]
Thanks for that information and your suggestion. Indeed I have always run with a single vcf containing both the parents and offspring. I have now tried running with the same vcf as -P, but that did not help (I did not split the vcf, just called with the same VCF as you suggested). 

I did see that I actually also had text in my jobs stdout file (when it failed). Perhaps it is informative for you:

When I called with the VCF also as -P.
```
Making permanent edge forbidden (219, 220).
Making permanent edge forbidden (224, 225).
Making permanent edge forbidden (886, 887).
Making permanent edge forbidden (1207, 1208).
Making permanent edge forbidden (1211, 1212).
Making permanent edge forbidden (1392, 1393).
Instance is infeasible!
```

Previous call without -P
```
Making permanent edge forbidden (239, 240).
Making permanent edge forbidden (248, 249).
Making permanent edge forbidden (250, 251).
Assertion violated (Get != Set): 315 1 -nan -nan
Assertion violated (Get != Set): 315 8 -nan -nan
Assertion violated (Get != Set): 315 15 -nan -nan
Assertion violated (Get != Set): 315 22 -nan -nan
Assertion violated (Get != Set): 315 29 -nan -nan
Assertion violated (Get != Set): 315 36 -nan -nan
Assertion violated (Get != Set): 315 43 -nan -nan
Assertion violated (Get != Set): 315 50 -nan -nan
Assertion violated (Get != Set): 315 57 -nan -nan
Assertion violated (Get != Set): 315 64 -nan -nan
Assertion violated (Get != Set): 315 67 -nan -nan
Assertion violated (Get != Set): 315 73 -nan -nan
Assertion violated (Get != Set): 315 76 -nan -nan
Assertion violated (Get != Set): 315 79 -nan -nan
Assertion violated (Get != Set): 315 82 -nan -nan
Assertion violated (Get != Set): 315 85 -nan -nan
Assertion violated (Get != Set): 315 88 -nan -nan
Assertion violated (Get != Set): 315 91 -nan -nan
Assertion violated (Get != Set): 315 94 -nan -nan
Assertion violated (Get != Set): 315 97 -nan -nan
Assertion violated (Get != Set): 315 100 -nan -nan
Assertion violated (Get != Set): 315 103 -nan -nan
Assertion violated (Get != Set): 315 109 -nan -nan
Assertion violated (Get != Set): 315 112 -nan -nan
Assertion violated (Get != Set): 315 115 -nan -nan
Assertion violated (Get != Set): 315 118 -nan -nan
Assertion violated (Get != Set): 315 124 -nan -nan
Assertion violated (Get != Set): 315 127 -nan -nan
Assertion violated (Get != Set): 315 130 -nan -nan
Assertion violated (Get != Set): 315 142 -nan -nan
Assertion violated (Get != Set): 315 145 -nan -nan
Assertion violated (Get != Set): 315 154 -nan -nan
Assertion violated (Get != Set): 315 157 -nan -nan
Assertion violated (Get != Set): 315 160 -nan -nan
Assertion violated (Get != Set): 315 163 -nan -nan
Assertion violated (Get != Set): 315 166 -nan -nan
Assertion violated (Get != Set): 315 169 -nan -nan
Assertion violated (Get != Set): 315 172 -nan -nan
Assertion violated (Get != Set): 315 175 -nan -nan
Assertion violated (Get != Set): 315 178 -nan -nan
Assertion violated (Get != Set): 315 181 -nan -nan
Assertion violated (Get != Set): 315 184 -nan -nan
Assertion violated (Get != Set): 315 187 -nan -nan
Assertion violated (Get != Set): 315 190 -nan -nan
Assertion violated (Get != Set): 315 193 -nan -nan
Assertion violated (Get != Set): 315 196 -nan -nan
Assertion violated (Get != Set): 315 199 -nan -nan
Assertion violated (Get != Set): 315 202 -nan -nan
Assertion violated (Get != Set): 315 205 -nan -nan
Assertion violated (Get != Set): 315 220 -nan -nan
Assertion violated (Get != Set): 315 223 -nan -nan
Assertion violated (Get != Set): 315 226 -nan -nan
Assertion violated (Get != Set): 315 235 -nan -nan
Assertion violated (Get != Set): 315 238 -nan -nan
Assertion violated (Get != Set): 315 241 -nan -nan
Assertion violated (Get != Set): 315 244 -nan -nan
Assertion violated (Get != Set): 315 247 -nan -nan
Assertion violated (Get != Set): 315 253 -nan -nan
Assertion violated (Get != Set): 315 254 -nan -nan
Assertion violated (Get != Set): 315 259 -nan -nan
Assertion violated (Get != Set): 315 264 -nan -nan
Assertion violated (Get != Set): 315 265 -nan -nan
Assertion violated (Get != Set): 315 272 -nan -nan
Assertion violated (Get != Set): 315 273 -nan -nan
Assertion violated (Get != Set): 315 282 -nan -nan
Assertion violated (Get != Set): 315 283 -nan -nan
Assertion violated (Get != Set): 315 284 -nan -nan
Assertion violated (Get != Set): 315 285 -nan -nan
Assertion violated (Get != Set): 315 288 -nan -nan
Assertion violated (Get != Set): 315 289 -nan -nan
Assertion violated (Get != Set): 315 290 -nan -nan
Assertion violated (Get != Set): 315 291 -nan -nan
Assertion violated (Get != Set): 315 294 -nan -nan
Assertion violated (Get != Set): 315 295 -nan -nan
Assertion violated (Get != Set): 315 296 -nan -nan
Assertion violated (Get != Set): 315 308 -nan -nan
Assertion violated (Get != Set): 315 311 -nan -nan
Assertion violated (Get != Set): 315 312 -nan -nan
Assertion violated (Get != Set): 315 313 -nan -nan
Assertion violated (Get != Set): 315 314 -nan -nan
Assertion violated (Get != Set): 316 2 -nan -nan
Assertion violated (Get != Set): 316 9 -nan -nan
Assertion violated (Get != Set): 316 16 -nan -nan
Assertion violated (Get != Set): 316 23 -nan -nan
Assertion violated (Get != Set): 316 30 -nan -nan
Assertion violated (Get != Set): 316 37 -nan -nan
Assertion violated (Get != Set): 316 44 -nan -nan
Assertion violated (Get != Set): 316 51 -nan -nan
Assertion violated (Get != Set): 316 58 -nan -nan
Assertion violated (Get != Set): 316 68 -nan -nan
Assertion violated (Get != Set): 316 71 -nan -nan
Assertion violated (Get != Set): 316 74 -nan -nan
Assertion violated (Get != Set): 316 77 -nan -nan
Assertion violated (Get != Set): 316 80 -nan -nan
Assertion violated (Get != Set): 316 83 -nan -nan
Assertion violated (Get != Set): 316 86 -nan -nan
Assertion violated (Get != Set): 316 89 -nan -nan
Assertion violated (Get != Set): 316 92 -nan -nan
Assertion violated (Get != Set): 316 95 -nan -nan
Assertion violated (Get != Set): 316 98 -nan -nan
Assertion violated (Get != Set): 316 101 -nan -nan
Assertion violated (Get != Set): 316 104 -nan -nan
Assertion violated (Get != Set): 316 107 -nan -nan
Assertion violated (Get != Set): 316 110 -nan -nan
Assertion violated (Get != Set): 316 113 -nan -nan
Assertion violated (Get != Set): 316 116 -nan -nan
Assertion violated (Get != Set): 316 119 -nan -nan
Assertion violated (Get != Set): 316 122 -nan -nan
Assertion violated (Get != Set): 316 125 -nan -nan
Assertion violated (Get != Set): 316 131 -nan -nan
Assertion violated (Get != Set): 316 146 -nan -nan
Assertion violated (Get != Set): 316 149 -nan -nan
Assertion violated (Get != Set): 316 155 -nan -nan
Assertion violated (Get != Set): 316 161 -nan -nan
Assertion violated (Get != Set): 316 164 -nan -nan
Assertion violated (Get != Set): 316 167 -nan -nan
Assertion violated (Get != Set): 316 176 -nan -nan
Assertion violated (Get != Set): 316 179 -nan -nan
Assertion violated (Get != Set): 316 182 -nan -nan
Assertion violated (Get != Set): 316 185 -nan -nan
Assertion violated (Get != Set): 316 188 -nan -nan
Assertion violated (Get != Set): 316 191 -nan -nan
Assertion violated (Get != Set): 316 194 -nan -nan
Assertion violated (Get != Set): 316 197 -nan -nan
Assertion violated (Get != Set): 316 200 -nan -nan
Assertion violated (Get != Set): 316 221 -nan -nan
Assertion violated (Get != Set): 316 224 -nan -nan
Assertion violated (Get != Set): 316 227 -nan -nan
Assertion violated (Get != Set): 316 236 -nan -nan
Assertion violated (Get != Set): 316 242 -nan -nan
Assertion violated (Get != Set): 316 245 -nan -nan
Assertion violated (Get != Set): 316 254 -nan -nan
Assertion violated (Get != Set): 316 259 -nan -nan
Assertion violated (Get != Set): 316 264 -nan -nan
Assertion violated (Get != Set): 316 265 -nan -nan
Assertion violated (Get != Set): 316 272 -nan -nan
Assertion violated (Get != Set): 316 273 -nan -nan
Assertion violated (Get != Set): 316 282 -nan -nan
Assertion violated (Get != Set): 316 283 -nan -nan
Assertion violated (Get != Set): 316 284 -nan -nan
Assertion violated (Get != Set): 316 285 -nan -nan
Assertion violated (Get != Set): 316 288 -nan -nan
Assertion violated (Get != Set): 316 289 -nan -nan
Assertion violated (Get != Set): 316 290 -nan -nan
Assertion violated (Get != Set): 316 291 -nan -nan
Assertion violated (Get != Set): 316 294 -nan -nan
Assertion violated (Get != Set): 316 295 -nan -nan
Assertion violated (Get != Set): 316 296 -nan -nan
Assertion violated (Get != Set): 316 308 -nan -nan
Assertion violated (Get != Set): 316 311 -nan -nan
Assertion violated (Get != Set): 316 312 -nan -nan
Assertion violated (Get != Set): 316 313 -nan -nan
Assertion violated (Get != Set): 316 314 -nan -nan
Instance is infeasible!
```

**Comment by schrins** [comment]
This is indeed helpful, in the sense that it shows where the error ist. The output is produced by the clustering algorithm, which as a result of the input being "unclusterable", returns an empty clustering. This then finally leads to the error message from your first post, because if the clustering is empty, there cannot be any maximum over the set of clusters.

For further debugging, I will have to reproduce these errors myself. If I am lucky, this is just some numeric issue (then `-nan` values already hint to this), if I am unlucky, there is some corner case not covered by the algorithm.

Thanks for providing all the output.

**Comment by mhooykaas** [comment]
Thank you for your efforts! If I can assist eg by sending you VCF file(s) that resulted in the error then please let me know.

**Comment by schrins** [comment]
I was able to do a few tests. I used the data from a region of chromosome 3, based on Solyntus v1.1 (StSOLv1.1ch03:60269000-60504000, the same as in the paper about the genetic polyphase module).

1. Running the original data and the "decluttered" data did not make a difference. The result on Altus_470bp is identical in the resulting VCFs.
2. I merged the two decluttered VCFs (parent and progreny). If I use the merged file as both parent-file and progeny file (via `-P`) the output is different from the VCFs in 1.)
3. If I run the merged file again without `-P`, the result is the same as with `-P`.

I have to investigate further what happens when the VCFs are merged. Merging the "non-decluttered" files gave me an error from bcftools, which I did not follow further. At least my decluttering and re-using the same file with and without `-P` did not change the results.

I did not run into any crashes during my tests. I was always using `--complexity-support 2`, because this seemed to be the more errornous setting. Did you use the same parental and progreny data as the paper? It looks similar, because you use the same reference genome and the same parent samples. If that is the case, you could tell me which region you tried to phase. Otherwise it would be nice, if you could share some VCFs and the pedigree file with me. Preferably the one from the first post, because this region looks quite short and is thus easier to debug.

**Comment by mhooykaas** [comment]
I am not sure if --complexity-support 2 is always more problematic, because I have mostly tested with default settings and often saw the error. It was informative though that for this particular VCF the error only occurred with the --complexity-support 2 setting.

I tried one other thing: splitting my vcf file into a parent and an offspring VCF file with vcfkeepsamples and using -P. However that resulted in an error at an earlier stage, at "Extracting progeny allele depths". Perhaps the VCFs generated this way are not correct.

I am indeed working on the same Altus/Colomba sequence data and used Solyntus as a reference, but the mapping parameters may have been slightly different and I have done the variant calling with Freebayes. (Also I have not downloaded the reads but used them as they are stored on our cluster so I am not sure if there could be a minor difference in e.g. filtering.)

Tests I have run:

Your VCF files on Zenodo, with parent and offspring VCF separate, all with default `complexity-support:`
- chr03 region VCFs -> no error
- chr04 region VCFs  -> no error
- chr05 region VCFs -> error -> this was surprising. Perhaps you can reproduce this?

My Freebayes VCF for both parent and offspring:
- same chr03 region as above, selected from my own Freebayes VCF -> no error
- same chr04 region as above, selected from my own Freebayes VCF -> no error
- whole chr03 Freebayes VCF: error
- few thousand markers from chr03 Freebayes VCF (more or less randomly dispersed over whole chromosome): no error with default and  --complexity-support 1, but error with  --complexity-support 2
- whole chr11 Freebayes VCF: error
- region of chr11 Freebayes VCF (StSOLv1.1ch11:5702203-11404404): no error for Colomba, then crashed during Altus cluster arrangement
- other region of chr11 Freebayes VCF: error

Shall I send you the "few thousand markers from chr03 Freebayes VCF" file for debugging via email (5 MB in size)? The StSOLv1.1ch11:5702203-11404404 region also failed (for Altus), but that takes longer to run.

**Comment by schrins** [comment]
For the chr05 region I can reproduce the error message. The reason is that there are no S/N-variants left for Colomba (Altus works without errors), because every variant that is homozygous on Altus was filtered out. However, S/S-variants are still left, so if you try to phase Colomba with complexity-support 1, it actually finds some phasable variants. It still crashes, though, because relationships between markers (i.e. whether they rather want to be on the same or on different haplotypes) are only computed between one S/N-markers and another marker (possibly a S/N-marker as well). So, without S/N-markers no relationships are computed and the clustering algorithm puts every marker in its own cluster, because that is the "optimal" solution.

Long story short: If no S/N-markers are found, whatshap should stop and provide a meaningful error message.

The example in your first post does contain S/N-markers, so there must be another problem. If you could send me the example (I assume this is the 5 MB file?), that would be great.

**Comment by schrins** [comment]
I created a new branch `issue-439` to fix the discovered problems. Basically, I changed the read routine for the pedigree file:
1. If a parent is specified as its own progeny, this line is ignored. Previously, the algorithm just continued and could compute garbage out of this.
2. If a progeny sample is specified in the pedigree file, but not available in the VCF, a warning will be shown and the entry will not be considered. Previously, the algorithm eventually threw some KeyError when accessing a dictionary.
3. It is now not possible anymore to specify more than one partner for any parent sample. While there could potentially cases where one might have this, it is not trivial to extend the algorithm in that way. I am not sure what the algorithm would have computed before, but it could not have been correct in all cases.

@mhooykaas 
The data you sent via mail should now work with the code changes, even with the two wrong trio entries in the pedigree file. Would it be possible for you to test all (or some) of the crashing runs on the changed code? You would have to clone the git repository, switch to the correct branch and compile Whatshap locally:

```
git clone https://github.com/whatshap/whatshap.git
git checkout issue-439
```
Depending on whether you used virtual environment or conda (or something else) you need to compile it, which requires some dependencies. For conda, you could for example do it like this:
```
conda create -n whtest python=3.10
pip install -e .
```
The `pip` routine takes care of all dependencies (we found that this works out better than to specifiy all dependencies in conda).

**Comment by mhooykaas** [comment]
Thank you for the changes. I will test the failing files on the new branch
(but it will take a couple of days).

<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virusvrij.www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

**Comment by schrins** [comment]
I had the same error a few days ago and could solve it by removing and recreating my conda environment. I thought it was just something messed up with my environment, because the version of the package `pulp` was the same as in a second environment that still worked fined. It seems like I have to investigate where this comes from. At least I can tell that in order to get to the new error, the place of the old error must have been successfully passed.

If the fix is not trivial, I will close this issue and create a new one for this specific problem, because it also affects other parts of the tool that use pulp.

**Comment by schrins** [comment]
I couldn't reproduce your error anymore, but it looks like the mentioned package `pulp` does not always have the same set of internal solvers preinstalled. Instead of using a hardcoded, fixed solver, I now poll available solvers and pick the one I like the most. Since I cannot test, whether the error still exists, I have to rely on you.

I updated the issue branch with the latest changes from the main branch, so if it breaks somewhere else, it likely comes from this intermediate update. It might be inconvenient, but if the changes in the main branch breaks the genetic phasing again, it is better we see this now than leaving it unfixed in the final version.

**Comment by schrins** [comment]
If the clustering runs for more than a day, it should be pretty safe to assume that it will never finish due to some error. The "Assertion violated (Get != Set): 190787 190782 -nan -nan" messages indicate that the input to the internal clustering algorithm contained invalid values, which makes the behavior compleltey undefined.

Thre should be no changes to the phasing algorithm in the branch, besides the pedigree handling/detection. The AssertionError you got for the branch is indeed new and should prevent the problem that the clustering receives any NaN-values. If the assertion was removed, it is likely that the branch-version will also be stuck forever on the clustering. I added the assertion, because you reported these "Assertion violated (Get != Set):" messages before. However, after preventing that samples can be treated as their own parents (which is still possible in the main version), these messages did not occur anymore and I thought the problem was fixed. Well, apprently not ...

I need to look closer to find out where these NaN-values come from. I will come back to this later.

**Comment by schrins** [comment]
I found a potential source for the NaN values, but I don't know if that was already the issue. Can you pull the changes, run `pip install -e .` (<- necessary this time!) and test again? It should be sufficient to just run Altus (with `--sample Altus_470bp`) to save some time.

**Comment by mhooykaas** [comment]
Thanks for the changes! After pulling the changes it works fine on a small VCF file, on Altus. On the entire VCF file it is still running on --complexity-support 2, but with --complexity-support 0 and --complexity-support 1 I did not get errors. 
(I do still have an open question regarding the regenotyping, but I will open a new thread as I assume that is unrelated to this issue.)

**Comment by schrins** [comment]
If the remaining runs were successful, please let me know. Then I will close this issue and merge the changes into the main branch.

**Comment by mhooykaas** [comment]
The run on chromosome 3 with --complexity-support 2 also finished without errors.

**Comment by schrins** [comment]
Great, thanks.
